src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -342,6 +344,25 @@ float irr_glsl_beckmann(in float a2, in float NdotH2)
     return irr_glsl_RECIPROCAL_PI * nom/denom;
 }
 
+//TODO this is taking into account beckmann NDF only, i'd like to also account for masking/shadowing term (visibility of microfacets)
+BSDFSample irr_glsl_beckmann_cos_gen_sample(in uvec2 _sample, in float _a2)
+{
+    vec2 u = vec2(_sample)/float(4294967295);
+    float phi = 2.0*irr_glsl_PI*u.x;
```

Comment:
Make a note that

http://www.pbr-book.org/3ed-2018/Light_Transport_I_Surface_Reflection/Sampling_Reflection_Functions.html#fragbit-1215
+
https://github.com/mmp/pbrt-v3/blob/9f717d847a807793fa966cf0eaa366852efef167/src/core/microfacet.cpp#L120

Has a sampling according to visible microfacets which is much shorter and simpler

Its a bit disgusting for implementation in GLSL

Reply:
yeah, i've seen this code and thats exactly the reason why i wanted to look for some other source before i'm going to copy PBRT's code. I have Heitz's paper about sampling ggx (ggx actually got 2nd paper with preetier and faster solution) and beckmann VNDF and possibly there i'll find something better